src/components/tl/mlx5/alltoall/alltoall_coll.c

```diff
@@ -0,0 +1,873 @@
+    int                          seq_index       = task->alltoall.seq_index;
+    int                          flag;
+    ucc_tl_mlx5_rcache_region_t *send_ptr;
+    ucc_tl_mlx5_rcache_region_t *recv_ptr;
+
+    UCC_TL_MLX5_PROFILE_REQUEST_EVENT(task, "mlx5_alltoall_fanin_start", 0);
+    tl_debug(UCC_TASK_LIB(task), "register memory buffers");
+    coll_task->status       = UCC_INPROGRESS;
+    coll_task->super.status = UCC_INPROGRESS;
+
+    errno = 0;
+    if (UCC_OK !=
+        ucc_rcache_get(ctx->rcache, (void *)SCHEDULE_ARGS(task).src.info.buffer,
+                       task->alltoall.msg_size * UCC_TL_TEAM_SIZE(team),
+                       &reg_change_flag, (ucc_rcache_region_t **)&send_ptr)) {
+        tl_error(UCC_TASK_LIB(task),
+                 "Failed to register send_bf memory (errno=%d)", errno);
+        return UCC_ERR_NO_RESOURCE;
+    }
+    task->alltoall.send_rcache_region_p = send_ptr;
+
+    /* NOTE: we don't support alternating block_size for the same msg size - TODO
+       Will need to add the possibility of block_size change into consideration
+       when initializing the mkey_cache_flag */
+
+    flag =
+        (task->alltoall.msg_size == a2a->previous_msg_size[seq_index])
+            ? 0
+            : (UCC_MLX5_NEED_SEND_MKEY_UPDATE | UCC_MLX5_NEED_RECV_MKEY_UPDATE);
+
+    if (reg_change_flag || (task->alltoall.send_rcache_region_p->reg.mr->addr !=
+                            a2a->previous_send_address[seq_index])) {
+        flag |= UCC_MLX5_NEED_SEND_MKEY_UPDATE;
+    }
+    reg_change_flag = 0;
+    if (UCC_OK !=
+        ucc_rcache_get(ctx->rcache, (void *)SCHEDULE_ARGS(task).dst.info.buffer,
+                       task->alltoall.msg_size * UCC_TL_TEAM_SIZE(team),
+                       &reg_change_flag, (ucc_rcache_region_t **)&recv_ptr)) {
+        tl_error(UCC_TASK_LIB(task), "Failed to register receive_bf memory");
+        ucc_rcache_region_put(ctx->rcache,
+                              &task->alltoall.send_rcache_region_p->super);
+        return UCC_ERR_NO_RESOURCE;
+    }
+    task->alltoall.recv_rcache_region_p = recv_ptr;
+    if (reg_change_flag || (task->alltoall.recv_rcache_region_p->reg.mr->addr !=
+                            a2a->previous_recv_address[seq_index])) {
+        flag |= UCC_MLX5_NEED_RECV_MKEY_UPDATE;
+    }
+
+    tl_debug(UCC_TL_MLX5_TEAM_LIB(team), "fanin start");
+    /* start task if completion event received */
+    /* Start fanin */
+    ucc_tl_mlx5_get_my_ctrl(a2a, seq_index)->mkey_cache_flag = flag;
+    ucc_tl_mlx5_update_mkeys_entries(a2a, task, flag);
+
+    if (UCC_OK == ucc_tl_mlx5_node_fanin(team, task)) {
+        tl_debug(UCC_TL_MLX5_TEAM_LIB(team), "fanin complete");
+        coll_task->status = UCC_OK;
+        UCC_TL_MLX5_PROFILE_REQUEST_EVENT(task, "mlx5_alltoall_fanin_done", 0);
+        return ucc_task_complete(coll_task);
+    } else {
+        ucc_progress_enqueue(UCC_TL_CORE_CTX(team)->pq, coll_task);
+    }
+    return UCC_OK;
```

Comment:
no need for else, just
```
return ucc_progress_enqueue(UCC_TL_CORE_CTX(team)->pq, coll_task);
```

Reply:
I agree that the else is not necessary, but I still need to return ucc_status, whereas ucc_progress_enqueue is void